# Exercise 4.8: Grouping Data & Aggregating Variables

# 01. Importing libraries

In [2]:
# Import the libraries

import pandas as pd
import numpy as np
import os

# 02. Importing data

In [4]:
# Import the latest ords_prods_merged data

path = r'/Users/vijaykasturi/Library/CloudStorage/OneDrive-Personal/Data Analytics Course - CareerFoundry/Data Immersion/A4/Instacart Basket Analysis'

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_4.8.pkl'))

# 03. Task 4.8

In [6]:
# 4.8.2: Finding aggregated mean of order_number' for entire data frame

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 4.8.3: Analysis of the Results
# The means in the complete data frame are slightly higher than those in the subset.
# Department_id 21 & 20 are still the top selling departments

In [14]:
# 4.8.4: Creating a loyalty flag for existing customers using the transform() and loc() functions

# Note - The steps were already performed as part of the exercise and the updated data was imported for this task. Results shown below

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [20]:
# 4.8.5: Analysing spending habits of different customer types

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

# Analysis of results:
# Average prices of purchases by Loyal customers are lower than the other 2 customer types
# New customers tend to purchase higher priced products on average

In [22]:
# 4.8.6: Targeting customers by Spending pattern

# Aggregate using mean prices

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/73/ncnyv2zs7tlgp3vnng5pbtv00000gn/T/ipykernel_97463/1441561579.py:5: FutureWarning: The provided callable <function mean at 0x105de1300> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [24]:
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             0   2539329        1             1                   2   
2             0   2539329        1             1                   2   
3             0   2539329        1             1                   2   
4             0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  8                     7.0       14084                  2   
2                  8                     7.0       12427                  3   
3                  8                     7.0       26088                  4   
4                  8                     7.0       26405                  5   

   reordered  ... department_id  prices _merge    price_range_loc  \
0          0  ...             7     9.0   both  Mid-range product   
1          0  ...            16    12.5   both  Mid-range product   
2          0  ...            19     4.4   both  Low-range product   
3          0  ...            19     4.7   both  Low-range product   
4          0  ...            17     1.0   both  Low-range product   

      busiest_day  daily_traffic_label busiest_period_of_day max_order  \
0  Regularly busy         Regular days        Average orders        10   
1  Regularly busy         Regular days        Average orders        10   
2  Regularly busy         Regular days        Average orders        10   
3  Regularly busy         Regular days        Average orders        10   
4  Regularly busy         Regular days        Average orders        10   

   loyalty_flag avg_price  
0  New customer  6.367797  
1  New customer  6.367797  
2  New customer  6.367797  
3  New customer  6.367797  
4  New customer  6.367797  

[5 rows x 24 columns]

In [26]:
# Assign Spending Flag

ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High Spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low Spender'

In [28]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

# 4.8.6: There are 634,245 High Spenders; the rest are Low Spenders.

In [30]:
# 4.8.7: Determing Frequent Customers

# Aggregate using median of ordering frequency

ords_prods_merge['ord_freq'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/73/ncnyv2zs7tlgp3vnng5pbtv00000gn/T/ipykernel_97463/4214076244.py:5: FutureWarning: The provided callable <function median at 0x105f1c9a0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['ord_freq'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [32]:
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             0   2539329        1             1                   2   
2             0   2539329        1             1                   2   
3             0   2539329        1             1                   2   
4             0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  8                     7.0       14084                  2   
2                  8                     7.0       12427                  3   
3                  8                     7.0       26088                  4   
4                  8                     7.0       26405                  5   

   reordered  ... _merge    price_range_loc     busiest_day  \
0          0  ...   both  Mid-range product  Regularly busy   
1          0  ...   both  Mid-range product  Regularly busy   
2          0  ...   both  Low-range product  Regularly busy   
3          0  ...   both  Low-range product  Regularly busy   
4          0  ...   both  Low-range product  Regularly busy   

   daily_traffic_label  busiest_period_of_day  max_order  loyalty_flag  \
0         Regular days         Average orders         10  New customer   
1         Regular days         Average orders         10  New customer   
2         Regular days         Average orders         10  New customer   
3         Regular days         Average orders         10  New customer   
4         Regular days         Average orders         10  New customer   

  avg_price spending_flag ord_freq  
0  6.367797   Low Spender     20.0  
1  6.367797   Low Spender     20.0  
2  6.367797   Low Spender     20.0  
3  6.367797   Low Spender     20.0  
4  6.367797   Low Spender     20.0  

[5 rows x 26 columns]

In [42]:
# Assign Spending Flag

ords_prods_merge.loc[ords_prods_merge['ord_freq'] <= 10, 'frequency_flag'] = 'Frequent Customer'
ords_prods_merge.loc[(ords_prods_merge['ord_freq'] > 10) & (ords_prods_merge['ord_freq'] <= 20), 'frequency_flag'] = 'Regular Customer'
ords_prods_merge.loc[ords_prods_merge['ord_freq'] > 20, 'frequency_flag'] = 'Non-Frequent Customer'

In [46]:
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             0   2539329        1             1                   2   
2             0   2539329        1             1                   2   
3             0   2539329        1             1                   2   
4             0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  8                     7.0       14084                  2   
2                  8                     7.0       12427                  3   
3                  8                     7.0       26088                  4   
4                  8                     7.0       26405                  5   

   reordered  ...    price_range_loc     busiest_day daily_traffic_label  \
0          0  ...  Mid-range product  Regularly busy        Regular days   
1          0  ...  Mid-range product  Regularly busy        Regular days   
2          0  ...  Low-range product  Regularly busy        Regular days   
3          0  ...  Low-range product  Regularly busy        Regular days   
4          0  ...  Low-range product  Regularly busy        Regular days   

   busiest_period_of_day  max_order  loyalty_flag avg_price spending_flag  \
0         Average orders         10  New customer  6.367797   Low Spender   
1         Average orders         10  New customer  6.367797   Low Spender   
2         Average orders         10  New customer  6.367797   Low Spender   
3         Average orders         10  New customer  6.367797   Low Spender   
4         Average orders         10  New customer  6.367797   Low Spender   

  ord_freq    frequency_flag  
0     20.0  Regular Customer  
1     20.0  Regular Customer  
2     20.0  Regular Customer  
3     20.0  Regular Customer  
4     20.0  Regular Customer  

[5 rows x 27 columns]

In [44]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        22790523
Regular Customer          6927608
Non-Frequent Customer     2686728
Name: count, dtype: int64

# 4.8.7: There are 22790523 Frequent Customers.

# 04. Exporting data

In [48]:
# Export ords_prods_merge as pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_4.8_final.pkl'))

# End of Exercise 4.8